feat(vrl): warning on if-statement predicate that always/never succeeds

JeanMertz opened the pull request:
This PR adds two new compilation warnings to VRL, both related to if-statement predicates.

If a predicate is known to always resolve to true or false at runtime, the compiler will reject the predicate, as this is either (1) unnecessarily complicated code for the `true` case (e.g. wrapping the code in an if-statement adds nothing of value to the program, but does increase the cognitive burden on readers). 

And (2) it is likely a bug introduced by the operator, as a predicate that always resolves to `false` means the code in the block never resolves at runtime.

As you can see in the diagnostic fixes, the compiler doesn't yet detect cases such as `if true == true`, but those will be tackled in a separate PR.

JeanMertz committed:
feat(vrl): error on if-statement predicate that always/never succeeds

Signed-off-by: Jean Mertz <git@jeanmertz.com>

fuchsnj commented:
I disagree with this change. There are many reasons someone might want a "constant" condition with an if/else statement.
During development / debugging of a config, if you have a long if/else you might want to temporarily change that to `true` or `false` for testing.
The config might also be generated from some templating engine, or any arbitrary code. The result of this might produce `true` or `false` as a condition, and that should be perfectly valid. 
I could understand wanting to produce a _warning_, but it shouldn't outright be rejected.

JeanMertz commented:
> I could understand wanting to produce a _warning_, but it shouldn't be outright rejected.

While the goal of VRL is to prevent as much runtime problems as possible (and I consider "accidental dead code" to be one of those), I do agree that we'll likely want to distinguish between errors and warnings. We've discussed/mused about those differences before.

I'll see if I can sneak such a system in easily, and update this PR to emit a warning instead of a compilation error.

jszwedko commented:
Additionally, I could see users doing something like:

```
if ${ENV} == "production" {
  // do some things related to production
}
```

Granted, this could be made dynamic by using `get_env_var` rather than interpolation.

I could see us having a strict mode in the future that makes dead code a boot time error.

JeanMertz commented:
I've rebased this PR on top of https://github.com/vectordotdev/vector/pull/12412. The error has now been reduced to a warning.

Note that the test harness (for now) doesn't validate the return value of the program when a warning is emitted. I've moved this to a follow-up task in the "VRL compiler improvements" list here: https://github.com/vectordotdev/vector/issues/8258.

JeanMertz commented on lib/vrl/compiler/src/expression/predicate.rs:
```diff
@@ -20,10 +20,26 @@ pub struct Predicate {
+            .map(|expr| (expr.type_def(state), expr.as_value()))
+            .unwrap_or_else(|| (TypeDef::null(), None));
+
+        match value {
+            Some(Value::Boolean(true)) => {
+                return Err(Error {
+                    variant: ErrorVariant::AlwaysTrue,
+                    span,
+                })
+            }
+            Some(Value::Boolean(false)) => {
+                return Err(Error {
+                    variant: ErrorVariant::AlwaysFalse,
+                    span,
+                })
+            }
+            _ => {}
+        }
```
I just noticed that I need to do some more work here. After we introduced the concept of "warnings" in the compiler, I realized that, while this error gets turned into a warning, the rest of the predicate is no longer compiled, so you end up with an invalid program, even though the compiler succeeds in its compilation.
